packages/flutter/lib/src/material/icon_button.dart

```diff
@@ -288,6 +289,9 @@ class IconButton extends StatelessWidget {
   ///  * [Feedback] for providing platform-specific feedback to certain actions.
   final bool enableFeedback;
 
+  /// Box constraints for the button.
+  final BoxConstraints constraints;
```

Comment:
The doc should briefly explain what happens by default (if the constraints are not specified).  It would also help to include a (short) paragraph that explains the use case that motivated this PR, i.e. for icons that are smaller than 48x48 one must...

RaisedButton, FlatButton, OutlineButton, all unconditionally apply the visual density adjustment. We should probably do the same here.

Note also: perhaps the following idiom (from RawMaterialButton) should be factored into a VisualDensity.effectiveConstraints() method.
```
    final Offset densityAdjustment = widget.visualDensity.baseSizeAdjustment;
    final BoxConstraints effectiveConstraints = widget.constraints.copyWith(
      minWidth: widget.constraints.minWidth != null ? (widget.constraints.minWidth + densityAdjustment.dx).clamp(0.0, double.infinity) as double : null,
      minHeight: widget.constraints.minWidth != null ? (widget.constraints.minHeight + densityAdjustment.dy).clamp(0.0, double.infinity) as double : null,
    );
```

CC @gspencergoog

Reply:
Thank you very much for the comment! I have a few questions:

1. What is a situation where `constraints.minWidth` can be `null`?
    - The [documentation on `BoxConstraints`](https://api.flutter.dev/flutter/rendering/BoxConstraints-class.html) mentions that the the parameters must satisfy 0.0 ≤ minWidth ≤ maxWidth ≤ double.infinity and 0.0 ≤ minHeight ≤ maxHeight ≤ double.infinity. Does this mean that `constraints.minWidth` should never be `null`?
2. If the incoming `BoxConstraints` parameter specifies `maxWidth` and `maxHeight`, should they also comply with `VisualDensity` adjustment?

Reply:
Please see latest commit, which does the following:
1. Adds assertions to `BoxConstraints` so that parameters are always non-null.
2. `BoxConstraints` does not modify existing `maxWidth` and `maxHeight`.
3. Removed the testing for `!= null`.

Reply:
[just returned from the holiday break]

I agree that BoxConstraints shouldn't allow null parameters. Unforunately asserting that the parameters are non-null is a potentially breaking change and a bit subtle.

It would be safest to roll the BoxConstraints change through by itself (separate PR) assuming that it's not in fact a breaking change.

In this case, the constructor should assert `constraints == null || constraints.debugAssertIsValid()`. Likewise for the new visualDensity.effectiveConstraints() method. The MaterialButton class should include the same assert for its constraints parameter.

Reply:
Added more documentation.

Reply:
Hi Hans,

There seems to be an issue with the assertion.

I added the `constraints == null || constraints.debugAssertIsValid()` assertion to the constructor, but it is considered as "invalid constant value" because `BoxConstraints.debugAssertIsValid()` is not constant, but the constructor for `IconButton` is constant. Analysis is not passed.

Is there a way to make it work? I thought of a few ways to work around it:
1. Add the assertions to the relevant methods in `IconButton` and `RawMaterialButton` where the value of `constraints` is used. This does not verify the validity of constraints upon construction, but only each time when the constraints are used.
2. Add the assertions to the constructor for `BoxConstraints`. But this could be a breaking change, requiring all `BoxConstraints` instances to be valid.

Using `constraints.minWidth` in constructor does not seem to work either — it looks like `constraints.minWidth` is not a constant.

How should we handle this situation?

Thank you very much.

Regards,
Tianguang

Reply:
The asserts in the initializer lists for const constructors can only be applied to const expressions.  The debugIsValid() check in VisualDensity.effectiveConstraints() is sufficient, we don't need to check that the constraints are valid here.